### ISSUE TITLE
Make consolidateAfter configurable by disruption reason

### ISSUE THREAD

**Issue by yagonobre** [kubernetes-sigs/karpenter#1757]
### Description

**What problem are you trying to solve?**

We are trying to better control consolidation and avoid pods being moved around too frequently. To achieve this, we started using `consolidateAfter`. However, this has the side effect of keeping empty nodes in our cluster for longer, leading to additional costs.

It would be ideal if we had the option to configure consolidateAfter based on the disruption reason, allowing us better control over the process. We understand that the Karpenter API is now stable, but what we are proposing is to add an optional consolidateAfter configuration per disruption reason, which would take higher priority than the global setting.

Here is a api suggestion:

```code
spec:
  disruption:
    budgets:
    - nodes: "2%"
      schedule: "0 21 * * *"
      duration: 12h
      reasons:
      - Drifted
    - nodes: "100%"
      consolidateAfter: 15m
      reasons:
      - Empty
    - nodes: "2%"
      reasons:
      - Underutilized
    consolidateAfter: 12h
```

Her we should have `Drifted` and `Unterutilized` Nodes consolidated after 12h but Empty nodes after 15m.

**How important is this feature to you?**

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request

**Comment by steve-todorov** [comment]
Not stale.

 /remove-lifecycle stale

**Comment by jonathan-innis** [comment]
So, the original idea around consolidateAfter (conceptually) was that these nodes are kept around and not consolidated for some minimum amount of time to reduce the chance that we delete a node and then immediately have to spin-up a new node for a pod. It _does_ make sense to me that this problem is slightly different for empty nodes although, I will say that there were two original considerations for too aggressive consolidation that we were trying to avoid: 1) Nodes are consolidated and then are immediately spun back-up again because a pod comes in right after the consolidation decision is made (this would apply to both empty and underutilized nodes and was the original reason why this feature was implemented in this way) and 2) We wanted to make sure that we didn't introduce too much pod churn for nodes that did already have pods scheduled -- this obviously isn't affected by empty nodes.

Are you concerned at all with the first callout?

**Comment by jonathan-innis** [comment]
Are you using the value `12h` for `consolidateAfter` that you have specified above?

**Comment by yagonobre** [comment]
> Are you concerned at all with the first callout?

I agree that this can be a problem, but our initial concern was just to reduce pod churn.

> Are you using the value 12h for consolidateAfter that you have specified above -- I'm curious why you're considering using a consolidateAfter value that is so long 

We were trying to keep pods running for at least 12 hours, but we are using 1 hour to avoid having too many empty nodes in the cluster.

> do you expect to have pods that don't schedule to a node within a 12h timeframe?

I would say yes because the average lifetime of a pod is probably longer than spot instance, but I need to double check our metrics to be sure.


Also, if there is another way to make consolidation take the pod's age into account, we are open to use it as well.

**Comment by jonathan-innis** [comment]
> Also, if there is another way to make consolidation take the pod's age into account

Yeah, it's a bit tricky because what you really want is to attach some minimum lifetime to pods that schedule to nodes rather than keep a node around when there has been churn around it for some amount of time -- the difference here as well is that pods that are evicted from the node also have the potential to extend the consolidateAfter (this is the case in emptiness when you have a pod that goes empty, but then you have to wait some extended amount of time to really know that you want to pull the node down) -- are there very specific pods that you want to extend the lifetime of while consolidating? I'm mainly questioning whether using a user-configurable consolidateAfter or pod lifetime is the right mechanism to reduce pod churn or if there's something smarter that Karpenter could do as it's taking consolidation decisions to reduce this as well.

**Comment by yagonobre** [comment]
Yes! We want to define a minimum lifetime for pods. In our use case, we want to apply this to all pods, but it would be nice to have the ability to define it at the pod level, so we can have different lifetimes depending on the application.

**Comment by makarov-roman** [comment]
I have a similar experience. 
In the cluster with a 20 spot nodes, none of them survive 1h with `consolidate_after: 10m`. Which leads to the constant rescheduling of everything. Now I'm using `consolidate_after: 1h`, just to increase the nodes lifespan. 
It feels like a workaround though, because I want to reduce aggressiveness of the consolidation process, and I don't really want to depend on a pods scheduling.

**Comment by camaeel** [comment]
I ended up with narrowing down consolidation window with disruptionBudgetes. So nodes can consolidate out of peak hours and only 1 or 2 at a time.
